src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs

```diff
@@ -1097,17 +1095,26 @@ public override void NotifyEndApplication()
         {
             lock (hostGlobalLock)
             {
-                Dbg.Assert(_beginApplicationNotifyCount > 0, "Not running an executable - NotifyBeginApplication was not called!");
-                --_beginApplicationNotifyCount;
-                if (_beginApplicationNotifyCount == 0)
+                if (--_beginApplicationNotifyCount == 0)
                 {
-                    // restore the window title when the last application started has ended.
-
+                    // Restore the window title when the last application started has ended.
                     ui.RawUI.WindowTitle = _savedWindowTitle;
 #if !UNIX
                     if (_savedConsoleMode != ConsoleControl.ConsoleModes.Unknown)
                     {
                         ConsoleControl.SetMode(ConsoleControl.GetActiveScreenBufferHandle(), _savedConsoleMode);
+                        if (_savedConsoleMode.HasFlag(ConsoleControl.ConsoleModes.VirtualTerminal))
+                        {
+                            // If the console output mode we just set already has 'VirtualTerminal' turned on,
+                            // we don't need to try turn on the VT mode separately.
+                            return;
+                        }
+                    }
+
+                    if (ui.SupportsVirtualTerminal)
+                    {
+                        // Re-enable VT mode if it was previously enabled, as a native command may have turned it off.
+                        ui.TryTurnOnVirtualTerminal();
```

Comment:
Nit: Interesting, is such console operations expensive?
In line 1105 we have already requested a handle and set mode, now TryTurnOnVirtualTerminal does the operations again. It seems we could do perf optimization.

Reply:
Very likely not expensive, especially if this is done only after native command execution. The handle is already cached, `GetActiveScreenBufferHandle()` returns a cached output handle.